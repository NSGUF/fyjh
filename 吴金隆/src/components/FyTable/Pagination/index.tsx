/**
 * @file 表格的分页器
*/
import { defineComponent } from "vue";
import './index.scss'

// TODO 每页显示数量可选

const SLIDE_WINDOW_WIDTH_5 = 5;
const SLIDE_WINDOW_WIDTH_5_MIDDLE = Math.ceil(SLIDE_WINDOW_WIDTH_5 / 2);

export default defineComponent({
  name: "Pagination",

  props: {
    total: {
      type: Number,
      required: true
    },
    change: {
      type: Function,
      default: () => {}
    },
    pageSize: {
      type: Array,
      default: []
    }
  },

  emits: ['change', 'pageSizeChange'],

  data() {
    return {
      curPage: 1,
      slideWindow: {
        leftMostPage: 1,
        rightMostPage: SLIDE_WINDOW_WIDTH_5,
        isAlive: true
      },
      curPageSize: this.pageSize[0],
      thePageJumpTo: 0,
    };
  },

  computed: {
    totalPage() {
      return Math.ceil(this.total / (this.curPageSize as number));
    }
  },

  watch: {
    curPage (curPage) {
      this.$emit('change', curPage);
    },

    curPageSize (pageSize) {
      this.slideWindow.isAlive = this.totalPage > SLIDE_WINDOW_WIDTH_5 + 2;

      this.$emit('pageSizeChange', pageSize);
    }
  },

  mounted() {
    this.slideWindow.isAlive = this.totalPage > SLIDE_WINDOW_WIDTH_5 + 2;
  },

  methods: {
    // 点击页码事件处理
    handleJumpTo(pageNum: number) {
      const { curPage, totalPage } = this;

      // 更新slideWindow
      if (this.slideWindow.isAlive) {
        this.moveSlideWindowTo(pageNum)
      }

      this.curPage = pageNum;
    },

    // 往前一页处理
    handleJumpPrev() {
      this.curPage = this.curPage > 1 ? this.curPage - 1 : 1;
      this.moveSlideWindowTo(this.curPage);
    },

    // 往后一页处理
    handleJumpNext() {
      this.curPage = this.curPage < this.totalPage ? this.curPage + 1 : this.totalPage;
      this.moveSlideWindowTo(this.curPage);
    },

    // 快速往前跳转处理
    handleJumpPrevQuickly () {
      let moveTo = this.curPage - SLIDE_WINDOW_WIDTH_5;

      this.curPage = moveTo < 1 ? 1 : moveTo;
      this.moveSlideWindowTo(moveTo);
    },

    // 快速往后跳转处理
    handleJumpNextQuickly () {
      let moveTo = this.curPage + SLIDE_WINDOW_WIDTH_5;

      this.curPage = moveTo > this.totalPage ? this.totalPage : moveTo;
      this.moveSlideWindowTo(moveTo);
    },

    /**
     * 移动slideWindow到目标页码，如果slidWindow已经到达两端，则停止滑动。
     * @param {Number} moveTo - 移动到的目标页。
     */
    moveSlideWindowTo (moveTo:number) {
      let leftMostPage = 0,
          rightMostPage = 0;

      if (moveTo <= SLIDE_WINDOW_WIDTH_5_MIDDLE) {
        leftMostPage = 1;
        rightMostPage = SLIDE_WINDOW_WIDTH_5;
      } else if (moveTo > this.totalPage - SLIDE_WINDOW_WIDTH_5_MIDDLE) {
        leftMostPage = this.totalPage - SLIDE_WINDOW_WIDTH_5 + 1;
        rightMostPage = this.totalPage;
      } else {
        leftMostPage = moveTo - SLIDE_WINDOW_WIDTH_5_MIDDLE + 1;
        rightMostPage = moveTo + SLIDE_WINDOW_WIDTH_5_MIDDLE - 1;
      }

      this.slideWindow.leftMostPage = leftMostPage;
      this.slideWindow.rightMostPage = rightMostPage;
    },

    handlePageSizeChange (e) {
      this.handleJumpTo(1);

      this.curPageSize = +e.target.value;
    },

    handlePageInput (e) {
      let jumpTo = e.target.value;
      let isDecimal = jumpTo.indexOf(".") !== -1;
      jumpTo = +jumpTo;

      if (0 < jumpTo && jumpTo <= this.totalPage && !isDecimal) {
        this.thePageJumpTo = jumpTo;

        this.handleJumpTo(jumpTo);
      } else {
        this.thePageJumpTo = 0;
      }
    }
  },

  render() {
    return this.total > 0 ? (
      <div class="t-right">
        <ul class="pagination-box">
          <li class="pagination-prev">
            <button disabled={this.curPage <= 1} onClick={this.handleJumpPrev}>&lt;</button>
          </li>
          {
            this.slideWindow.isAlive ?
              <>
                { this.slideWindow.leftMostPage !== 1 ?
                  <>
                    <li class="pagination-item" onClick={() => this.handleJumpTo(1)}>1</li>
                    { this.curPage === SLIDE_WINDOW_WIDTH_5_MIDDLE + 1 ? null : <li onClick={this.handleJumpPrevQuickly} class="pagination-jump-prev" title="往前5页">...</li> }
                  </> : null
                }
                {
                  new Array(SLIDE_WINDOW_WIDTH_5).fill(this.slideWindow.leftMostPage).map((left, i) => (
                    <li
                      class={this.curPage === left + i ? 'pagination-item pagination-item-active' : 'pagination-item'}
                      onClick={() => this.handleJumpTo(left + i)}
                    >
                      {left + i}
                    </li>
                  ))
                }
                { this.slideWindow.rightMostPage !== this.totalPage ?
                  <>
                    { this.curPage === this.totalPage - SLIDE_WINDOW_WIDTH_5_MIDDLE ? null : <li onClick={this.handleJumpNextQuickly} class="pagination-jump-next" title="往后5页">...</li> }
                    <li class="pagination-item" onClick={() => this.handleJumpTo(this.totalPage)}>{this.totalPage}</li>
                  </> : null
                }
              </> :
              <>
                {
                  new Array(this.totalPage).fill('').map((item, i) => (
                    <li
                      class={this.curPage === i + 1 ? 'pagination-item pagination-item-active' : 'pagination-item'}
                      onClick={() => this.handleJumpTo(i + 1)}
                    >
                      {i + 1}
                    </li>
                  ))
                }
              </>
          }
          <li class="pagination-next">
            <button disabled={this.curPage === this.totalPage} onClick={this.handleJumpNext}>&gt;</button>
          </li>
        </ul>

        <select class="page-size-select global-mg-l-8" onChange={this.handlePageSizeChange}>
          {this.pageSize.map((size) => <option value={size}>{size}</option>)}
        </select>

        <span class="global-mg-l-8 global-mg-r-8">跳至</span>

        <input
          class="page-input global-pd-5"
          type="number"
          min="1"
          max={this.totalPage}
          value={this.thePageJumpTo || null}
          onBlur={this.handlePageInput}
        />
      </div>
    )
      : null;
  }
});
