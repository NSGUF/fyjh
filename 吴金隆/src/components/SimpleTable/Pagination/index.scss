.pagination-box {
  list-style: none;
  text-align: right;
}

.pagination-box .pagination-item {
  box-sizing: border-box;
  display: inline-block;
  width: 32px;
  height: 32px;
  line-height: 32px;
  margin-right: 8px;
  text-align: center;
  color: #000000d9;
  cursor: pointer;
  transition: all .3s;
  vertical-align: middle;
  border-radius: 2px;
  outline: 0;
  list-style: none;
  user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  border: 1px solid #d9d9d9;
}

.pagination-prev {
  margin-right: 8px;
}

.pagination-box .pagination-prev-disable,.pagination-box .pagination-next-disable {
  cursor: not-allowed;
  color: #d9d9d9;
}

.pagination-prev, .pagination-next {
  display: inline-block;
  width: 32px;
  height: 32px;
  line-height: 32px;
  text-align: center;
  color: #000000d9;
  vertical-align: middle;
  user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  outline: 0;

}

.pagination-prev button, .pagination-next button {
  display: block;
  width: 100%;
  height: 100%;
  background-color: #fff;
  border: 1px solid #d9d9d9;
  border-radius: 2px;
  transition: all .3s;
  cursor: pointer;

  &:disabled {
    cursor: not-allowed;
  }
  &:not(:disabled):hover {
    border-color: $global-primary-blue;
    color: $global-primary-blue;
  }
}

.pagination-item.pagination-item-active, .pagination-item:hover{
  font-weight: 500;
  border-color: $global-primary-blue;
  color: $global-primary-blue;
}