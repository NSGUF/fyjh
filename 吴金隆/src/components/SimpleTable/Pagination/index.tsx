/**
 * @file 表格的分页器
*/
import { defineComponent, ref, reactive } from "vue";
import './index.scss'

export default defineComponent({
    name: "Pagination",
  
    props: {
      total: {
        type: Number,
        required: true
      },
      change: {
        type: Function,
        default: () => {}
      },
      pageSize: {
        type: Number,
        default: 10
      }
    },

    data() {
      return {
        curPage: 1
      };
    },

    computed: {
      startPage () {
        return !!this.total ? 1 : 0;
      },

      endPage () {
        return this.total;
      },

      totalPage () {
        return Math.ceil(this.total/this.pageSize);
      }
    },

    methods: {
      handleClickItem ( pageNum: number) {
        this.curPage = pageNum;
        this.change(pageNum);
      },

      handleJumpPrev () {
        this.curPage = this.curPage > 1 ? this.curPage - 1 : 1;
        this.change(this.curPage);
      },

      handleJumpNext () {
        this.curPage = this.curPage < this.totalPage ? this.curPage + 1 : this.totalPage;
        this.change(this.curPage);
      }
    },

    render() {
      return this.total !== 0 ? (
        <ul class="pagination-box">
          <li class="pagination-prev">
            <button disabled={this.curPage <= 1} onClick={this.handleJumpPrev}>&lt;</button>
          </li>
          {/* 动态生成pagination-item */}
          {
            new Array(this.totalPage).fill('').map((item, i) => {
              return <li
                      class={this.curPage === i + 1 ? 'pagination-item pagination-item-active' : 'pagination-item'}
                      onClick={() => this.handleClickItem(i + 1)}
                      >
                      {i + 1}
                    </li>
            })
          }
          <li class="pagination-next">
            <button disabled={this.curPage === this.totalPage} onClick={this.handleJumpNext}>&gt;</button>
          </li>
        </ul>
      )
      : null;
    }
 });
 