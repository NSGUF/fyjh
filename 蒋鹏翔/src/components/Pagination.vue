<template>
    <div class="pagination">
        <button :disabled="pageNum === 1" @click="pageCompute(pageNum - 1)">
            上一页
        </button>
        <button v-if="startEnd.start !== 1" @click="pageCompute(1)">1</button>
        <button v-if="startEnd.start > 2">···</button>

        <span v-for="page in startEnd.end" :key="page">
            <button
                :class="{ active: pageNum === page }"
                v-if="page >= startEnd.start"
                @click="pageCompute(page)"
            >
                {{ page }}
            </button>
        </span>

        <button v-if="startEnd.end < totalPageNo - 1">···</button>
        <button
            v-if="startEnd.end !== totalPageNo"
            @click="pageCompute(totalPageNo)"
        >
            {{ totalPageNo }}
        </button>
        <button
            :disabled="pageNum === totalPageNo"
            @click="pageCompute(pageNum + 1)"
        >
            下一页
        </button>
        <input
            v-model.number="val"
            type="text"
            class="page-input"
            @blur="inputPage"
        />
        <button style="margin-left: 30px">共 {{ total }} 条</button>
    </div>
</template>

<script>
export default {
    name: "JPagination",

    props: {
        //   当前页
        pageNum: {
            type: Number,
            default: 1,
        },
        // 总数量
        total: {
            type: Number,
            default: 1,
        },
        // 每页数量
        pageSize: {
            type: Number,
            default: 3,
        },
        // 连续页数
        continueNo: {
            type: Number,
            default: 3,
        },

        changePageNo: {
            type: Function,
        },
    },
    data: () => {
        return {
            val: 1,
        };
    },
    computed: {
        //计算总页数
        totalPageNo() {
            return Math.ceil(this.total / this.pageSize);
        },

        startEnd() {
            let start = 0;
            let end = 0;
            let { totalPageNo, pageNum, continueNo } = this;

            if (totalPageNo <= continueNo) {
                start = 1;
                end = totalPageNo;
            } else {
                start = pageNum - Math.floor(continueNo / 2);
                end = pageNum + Math.floor(continueNo / 2);
                if (start <= 1) {
                    start = 1;
                    end = continueNo;
                }
                if (end >= totalPageNo) {
                    end = totalPageNo;
                    start = end - continueNo + 1;
                }
            }

            return { start, end };
        },
    },

    methods: {
        pageCompute(num) {
            this.val = num
            this.changePageNo(Math.abs(num));
        },

        inputPage() {
            let num = this.val;
            if (typeof num === "number") {
                num =
                    num > this.totalPageNo
                        ? this.totalPageNo
                        : num < 1
                        ? 1
                        : num;
                this.val = num;
                this.changePageNo(Math.abs(num));
            } else {
                this.val = 1;
            }
        },
    },
};
</script>

<style lang="less" scoped>
.pagination {
    button {
        margin: 0 5px;
        background-color: #f4f4f5;
        color: #606266;
        outline: none;
        border-radius: 2px;
        padding: 0 4px;
        vertical-align: top;
        display: inline-block;
        font-size: 20px;
        min-width: 35.5px;
        height: 28px;
        line-height: 28px;
        cursor: pointer;
        box-sizing: border-box;
        text-align: center;
        border: 0;

        &[disabled] {
            color: #c0c4cc;
            cursor: not-allowed;
        }

        &.active {
            cursor: not-allowed;
            background-color: #409eff;
            color: #fff;
        }
    }
    .page-input {
        width: 20px;
        height: 22px;
    }
}
</style>
