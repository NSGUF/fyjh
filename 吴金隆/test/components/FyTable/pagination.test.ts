import { mount } from "@vue/test-utils";
import { expect, test } from "vitest";
import Pagination from "../../../src/components/FyTable/Pagination";

test('dataLength === 10 时，测试分页器，只有一页，默认位于首页，不可左跳，不可右跳', () => {
  const dataLength = 10;
  const pageSize = [10];
  const wrapper = mount(Pagination, {
    props: {
      pageSize,
      total: dataLength,
    }
  });

  expect(wrapper.find('.pagination-item-active').text()).toBe('1');

  expect(wrapper.find('.pagination-box .pagination-prev button').attributes('disabled')).not.toBe('undefined');
  expect(wrapper.find('.pagination-box .pagination-next button').attributes('disabled')).not.toBe('undefined');
});

test('dataLength === 200 时，测试分页器，快进快退5页', async () => {
  const dataLength = 200;
  const pageSize = [10];
  const wrapper = mount(Pagination, {
    props: {
      pageSize,
      total: dataLength
    }
  });

  let curPage = 1;
  let getCurPage = () => {
    return +wrapper.find('.pagination-item-active').text();
  };
  let prevBtn = wrapper.find('.pagination-prev button');
  let nextBtn = wrapper.find('.pagination-next button');
  let jumpPrevBtn = wrapper.find('.pagination-jump-prev');
  let jumpNextBtn = wrapper.find('.pagination-jump-next');
  const paginationBoxEle = wrapper.find('.pagination-box');

  expect(paginationBoxEle.findAll('.pagination-item').length).toBe(6);

  // curPage = 1 时，无法左移
  await prevBtn.trigger('click');
  expect(getCurPage()).toBe(curPage);

  // 右移一页
  await nextBtn.trigger('click');
  curPage++;
  expect(getCurPage()).toBe(curPage);

  // 左移一页
  await prevBtn.trigger('click');
  curPage--;
  expect(getCurPage()).toBe(curPage);

  // 初始化不显示快速前进按钮
  expect(jumpPrevBtn.exists()).toBeFalsy();

  await jumpNextBtn.trigger('click');
  curPage += 5;
  expect(getCurPage()).toBe(curPage);

  // 回到初始状态
  jumpPrevBtn = wrapper.find('.pagination-jump-prev');
  await jumpPrevBtn.trigger('click');
  curPage = 1;
  expect(getCurPage()).toBe(curPage);

  let the20PageItem = wrapper.findAll('.pagination-box .pagination-item')[5];
  await the20PageItem.trigger('click');
  curPage = 20;
  expect(getCurPage()).toBe(curPage);
});

test('dataLength = 0 时，', () => {
  const dataLength = 0;
  const pageSize = [10];
  const wrapper = mount(Pagination, {
    props: {
      pageSize,
      total: dataLength,
    }
  });

  expect(wrapper.html()).toBeFalsy();
});


test('当dataLength = 120时，改变pageSize', async () => {
  const dataLength = 120;
  const pageSize = [10, 20, 30];
  const wrapper = mount(Pagination, {
    props: {
      pageSize,
      total: dataLength,
    }
  });
  const paginationItems = wrapper.findAll('.pagination-box .pagination-item');
  const pageSizeSelectEle = wrapper.find('.page-size-select');
  let thePageSize = '30';

  await pageSizeSelectEle.setValue(thePageSize);
  expect(paginationItems[paginationItems.length - 1].text()).toBe(120/30);
});

test('当dataLength = 120时，测试输入跳转', async () => {
  const dataLength = 120;
  const pageSize = [10, 20, 30];
  const wrapper = mount(Pagination, {
    props: {
      pageSize,
      total: dataLength,
    }
  });
  const pageInputEle = wrapper.find('.page-input');
  let jumpTo = '2'

  // 输入2
  await pageInputEle.setValue(jumpTo);
  await pageInputEle.trigger('blur');
  expect(wrapper.find('.pagination-item-active').text()).toBe(jumpTo);

  // 输入中文，无跳转响应
  let ChineseText = '中文测试';

  await pageInputEle.setValue(ChineseText);
  await pageInputEle.trigger('blur');
  expect(wrapper.find('.pagination-item-active').text()).toBe(jumpTo);

  // 输入100000000000
  let largeNumber = 100000000000;

  await pageInputEle.setValue(largeNumber);
  await pageInputEle.trigger('blur');
  expect(wrapper.find('.pagination-item-active').text()).toBe(jumpTo);
});