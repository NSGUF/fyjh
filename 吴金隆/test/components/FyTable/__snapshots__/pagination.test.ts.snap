// Vitest Snapshot v1

exports[`dataLength === 200 时，测试分页器，快进快退5页 1`] = `
"<ul class=\\"pagination-box\\">
  <li class=\\"pagination-prev\\"><button disabled=\\"\\">&lt;</button></li>
  <!---->
  <li class=\\"pagination-item pagination-item-active\\">1</li>
  <li class=\\"pagination-item\\">2</li>
  <li class=\\"pagination-item\\">3</li>
  <li class=\\"pagination-item\\">4</li>
  <li class=\\"pagination-item\\">5</li>
  <li class=\\"pagination-jump-next\\" title=\\"往后5页\\">...</li>
  <li class=\\"pagination-item\\">20</li>
  <li class=\\"pagination-next\\"><button>&gt;</button></li>
</ul>"
`;
